fuseutil/dirent.go

```diff
@@ -103,3 +117,136 @@ func WriteDirent(buf []byte, d Dirent) (n int) {
 
 	return n
 }
+
+// Write the supplied directory entry with attributes into the given buffer in the format
+// expected in fuseops.ReadDirPlusOp.Dst returning the number of bytes written.
+// Return zero if the entry would not fit.
+func WriteDirentPlus(buf []byte, d DirentPlus) (n int) {
+	type fuse_attr struct {
+		ino       uint64
+		size      uint64
+		blocks    uint64
+		atime     uint64
+		mtime     uint64
+		ctime     uint64
+		atimensec uint32
+		mtimensec uint32
+		ctimensec uint32
+		mode      uint32
+		nlink     uint32
+		uid       uint32
+		gid       uint32
+		rdev      uint32
+		blksize   uint32
+		padding   uint32
+	}
+
+	type fuse_entry_out struct {
+		nodeid           uint64
+		generation       uint64
+		entry_valid      uint64
+		attr_valid       uint64
+		entry_valid_nsec uint32
+		attr_valid_nsec  uint32
+		attr             fuse_attr
+	}
+
+	type fuse_dirent struct {
+		ino     uint64
+		off     uint64
+		namelen uint32
+		type_   uint32
+		name    [0]byte
+	}
+
+	// We want to write bytes with the layout of fuse_direntplus
+	// (http://shortn/_LNqd8uXg2p) in host order. The struct must be aligned
+	// according to FUSE_DIRENT_ALIGN (https://tinyurl.com/3m3ewu7h), which
+	// dictates 8-byte alignment.
+	type fuse_direntplus struct {
+		entry_out fuse_entry_out
+		dirent    fuse_dirent
+	}
+
+	const direntPlusAlignment = 8
+
+	//size of fuse_attr
+	const fuseAttrSize = 8 + 8 + 8 + 8 + 8 + 8 + 4 + 4 + 4 + 4 + 4 + 4 + 4 + 4 + 4 + 4
+
+	// size of fuse_entry_out without fuse_attr
+	const fuseEntryOutSize = 8 + 8 + 8 + 8 + 4 + 4
+
+	// size of fuse_dirent
+	const fuseDirentSize = 8 + 8 + 4 + 4
+
+	const direntPlusHeaderSize = fuseAttrSize + fuseEntryOutSize + fuseDirentSize
+
+	// Compute the number of bytes of padding we'll need to maintain alignment
+	// for the next entry.
+	var padLen int
+	if len(d.Dirent.Name)%direntPlusAlignment != 0 {
```

Comment:
Don’t compute the same expression twice, instead assign it to a variable in the if conditional:

```go
if pad := len(d.Dirent.Name)%direntPlusAlignment; pad != 0 {
	padLen = direntPlusAlignment - pad
}
```